### PR TITLE
feat(expr): support array_remove

### LINKED ISSUE #9114: implement array_remove
array_remove ( anycompatiblearray, anycompatible ) → anycompatiblearray

### PR DESCRIPTION
**Pull request by Honeta**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

array_remove ( anycompatiblearray, anycompatible ) → anycompatiblearray

Removes all elements equal to the given value from the array. Unlike standard pqsl, multidimensional arrays are supported in our implementation.

close #9114 

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

</details>

### CODE REVIEW

File: src/expr/src/expr/expr_array_remove.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    /// select array_remove(null::int[], 1);
+    /// ----
+    /// NULL
+    ///
+    /// query error unknown type
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], 1);
+    ///
+    /// query error unknown type
+    /// select array_remove(array[array[1],array[2],array[3],array[2],null::int[]], array[array[3]]);
+    /// ```
```
Comment (st1page) [inline src/expr/src/expr/expr_array_remove.rs:88]: please add some test with datatype's alignment

File: src/expr/src/expr/expr_array_remove.rs [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+                )
+                .into(),
+            ),
+            _ => None,
+        }
+    }
+
+    fn evaluate(&self, left: DatumRef<'_>, right: DatumRef<'_>) -> Datum {
+        Self::array_remove(left, right)
+    }
```
Comment (xiangjinwu) [inline src/expr/src/expr/expr_array_remove.rs:115]: Why do we need this?

File: src/expr/src/expr/expr_array_remove.rs [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    fn return_type(&self) -> DataType {
+        self.return_type.clone()
+    }
+
+    async fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let left_array = self.left.eval_checked(input).await?;
+        let right_array = self.right.eval_checked(input).await?;
+        let mut builder = self
+            .return_type
+            .create_array_builder(left_array.len() + right_array.len());
```
Comment (xiangjinwu) [inline src/expr/src/expr/expr_array_remove.rs:129]: This should be `input.capacity()` which should always be the same as `left_array.len()` and `right_array.len()`. We are wasting 1x reserved memory here. (And yes, `array_concat` also looks buggy here...)

**Review by xiangjinwu — APPROVED** [review]
Rest lgtm. Try removing `--@ ` for some of the tests below and run [risingwave_regress_test](https://github.com/risingwavelabs/risingwave/blob/main/src/tests/regress/README.md) to see if they can pass. It is okay to keep failing ones commented and fix them in a followup PR. I do see `align_array_and_element` is problematic when handling `unknown` (e.g. `null` and `'{1,2}'`).

https://github.com/risingwavelabs/risingwave/blob/57e4c259d9c90c302231c5019aa9df4ed5700229/src/tests/regress/data/sql/arrays.sql#L621-L627

File: src/expr/src/expr/expr_array_remove.rs (hunk 2 above)
Comment (xiangjinwu) [reply src/expr/src/expr/expr_array_remove.rs]: It is unnecessary to have `&self`.

**Pull request merged**